File: ocaml/01-10/09.ml
(* Pack consecutive duplicates of list elements into sublists. (medium) *)

let collect xs = List.fold_left 
  (fun (c, cs, all) x -> if x = c then (c, x::cs, all) else (x, [x], cs::all)) 
  (List.hd xs, [], []) 
  xs


(* pack : 'a list -> 'a list list *)
let pack = function 
  | []  -> []
  | [x] -> [[x]]
  | xs  -> let (_, ys, zs) = collect xs in List.rev (ys::zs)



let _ = assert (
  (pack ["a";"a";"a";"a";"b";"c";"c";"a";"a";"d";"d";"e";"e";"e";"e"]) =
  [["a"; "a"; "a"; "a"]; ["b"]; ["c"; "c"]; ["a"; "a"]; ["d"; "d"];
   ["e"; "e"; "e"; "e"]])

File: prolog/01-10/09.pl
/*
P09 (**) Pack consecutive duplicates of list elements into sublists.
    If a list contains repeated elements they should be placed in separate sublists.

    Example:
    ?- pack([a,a,a,a,b,c,c,a,a,d,e,e,e,e],X).
    X = [[a,a,a,a],[b],[c,c],[a,a],[d],[e,e,e,e]]
*/

pacc([], Ys, Acc) :- reverse(Acc, Ys).
pacc([H|T], Ys, [[H|Hs]|Acc]) :- pacc(T, Ys, [[H, H |Hs]|Acc]).
pacc([H|T], Ys, Acc) :- pacc(T, Ys, [[H]|Acc]).
pacc([H, G | T], Ys, Acc) :- 
  H \= G, 
  pacc([G|T], Ys, [[H]|Acc]).

pack(Xs, Ys) :- once(pacc(Xs, Ys, [])).
